lib/execution_segment.go

```diff
@@ -259,27 +260,23 @@ func roundUp(rat *big.Rat) *big.Int {
 
 // Scale proportionally scales the supplied value, according to the execution
 // segment's position and size of the work.
+// This implements the first proposal from #997:
+//   floor( (value * from) % 1 + value * length )
 func (es *ExecutionSegment) Scale(value int64) int64 {
 	if es == nil { // no execution segment, i.e. 100%
 		return value
 	}
-	// Instead of the first proposal that used remainders and floor:
-	//    floor( (value * from) % 1 + value * length )
-	// We're using an alternative approach with rounding that (hopefully) has
-	// the same properties, but it's simpler and has better precision:
-	//    round( (value * from) - round(value * from) + (value * (to - from)) )?
-	// which reduces to:
-	//    round( (value * to) - round(value * from) )?
-
-	toValue := big.NewRat(value, 1)
-	toValue.Mul(toValue, es.to)
-
 	fromValue := big.NewRat(value, 1)
 	fromValue.Mul(fromValue, es.from)
 
-	toValue.Sub(toValue, new(big.Rat).SetFrac(roundUp(fromValue), oneBigInt))
+	fv, _ := fromValue.Float64()
+	rem := math.Mod(fv, 1)
```

Comment:
I wasn't sure how to calculate `% 1` using `big.Rat`, i.e. without conversion to float. I don't think there should be a loss of precision because of this, but is there a simple way of doing this with rationals? Forgive my rusty math...

Reply:
hmm `%1` is just the remainder, so you should be able to calculate it by [dividing](https://golang.org/pkg/math/big/#Int.Rem) the [numerator](https://golang.org/pkg/math/big/#Rat.Num) of the rational number by its [denominator](https://golang.org/pkg/math/big/#Rat.Denom).

Reply:
and if you're only seeing a off-by-one error, there's a good chance it's caused by some float rounding shenanigans, so it's a good idea to convert this to only use rational numbers before we start investigating other potential causes

Reply:
:man_facepalming: Thanks... I experimented with `QuoRem`, but couldn't get it working.

It's fixed in latest push and the test now passes consistently for this implementation as well. :tada:

Reply:
:tada: those nasty floats :sweat_smile: 

'm very, very glad that I decided to stick with rational numbers - from experience, these types of off-by-one bugs are one of the most annoying ones to chase down. The Go rational numbers are a bit cumbersome to work with, and CPU-heavy, so we might switch to a lighter `int64`-based library in the future, but we're definitely sticking with the rational numbers to avoid issues like that :smile: